src/consensus/pool/finality_tracker.rs

```diff
@@ -0,0 +1,305 @@
+    ///
+    /// Handles possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn add_parent(&mut self, block: BlockId, parent: BlockId) -> FinalizationEvent {
+        let mut event = FinalizationEvent::default();
+        self.parents.insert(block, parent);
+        if let Some(status) = self.status.get(&block.0)
+            && let FinalizationStatus::FinalizedAndNotarized(block_hash)
+            | FinalizationStatus::ImplicitlyFinalized(block_hash) = status
+            && block.1 == *block_hash
+        {
+            self.handle_implicitly_finalized(block.0, parent, &mut event);
+        }
+        event
+    }
+
+    /// Marks the given block as fast finalized.
+    ///
+    /// If the block was newly finalized, handles resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_fast_finalized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        let mut event = FinalizationEvent::default();
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::FinalizedAndNotarized(_) => {
+                    return event;
+                }
+                FinalizationStatus::Notarized(_)
+                | FinalizationStatus::Finalized
+                | FinalizationStatus::ImplicitlyFinalized(_) => {}
+                FinalizationStatus::ImplicitlySkipped => unreachable!("consensus safety violation"),
+            }
+        }
+
+        self.status
+            .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+        self.handle_finalized((slot, block_hash), &mut event);
+        self.highest_finalized = slot.max(self.highest_finalized);
+        event
+    }
+
+    /// Marks the given block as notarized.
+    ///
+    /// Handles possibly resulting direct finalization of the block.
+    /// Further, also handles any possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_notarized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        let mut event = FinalizationEvent::default();
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::Notarized(_)
+                | FinalizationStatus::FinalizedAndNotarized(_)
+                | FinalizationStatus::ImplicitlyFinalized(_)
+                | FinalizationStatus::ImplicitlySkipped => event,
+                FinalizationStatus::Finalized => {
+                    self.status
+                        .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+                    self.handle_finalized((slot, block_hash), &mut event);
+                    event
```

Comment:
same comment as above.  I would define the event above this line and in the rest of the arms return `FinalizationEvent::default()`